### DISCUSSION TITLE
How to run it locally?

### DISCUSSION THREAD

**Discussion by autili** [fbarresi/SoftBeckhoff#4] [Q&A] [answered]
Hello Federico,

I could write to you in Italian, as I guess you are from Italy, but for letting more people understand I will use my basic English.

First of all many compliments for all your projects. They are all impressive. I will study each line of code of all them.

If I am not wrong what you are doing with SoftBeckhoff is a C# .NET ADS Server that will be able to expose variables through symbols.
I am sure you are aware that also Beckhoff has recently released a Nuget package (https://www.nuget.org/packages/Beckhoff.TwinCAT.Ads.Server/) that should be 'a base framework to create your own ADS server/virtual ADS device'.
I think that what you are trying to do with SoftBeckhoff is a 'virtual ADS device'.
The very interesting thing about your project is that you are going to expose symbols (and related types) to other ADS clients (just as the Beckhoff TwinCAT 3 PLC is able to do).
I asked directly to Beckhoff at the end of 2020 and they answered me that they have no plans to implement it into their package.
Through reverse engineering you are adding this functionality to your SoftBeckhoff.
And this is very interesting to me.

I have some skills in PLC programming and in 'classic' C# programming. I am a little late with my skills in the new .NET core flavour. I have no skills in web technologies (REST, containers, ASP .NET, ...): I am trying to fill the gap.

So, let's go to the question...
I am trying to understand the SoftBeckhoff code and I would like to test it locally using MS VS 2019.
The project compiles fine, but, when I run it, a page of the browser (MS Edge in my PC) opens up showing me the following error:
This localhost page can’t be found. No webpage was found for the web address: https://localhost:44305/. HTTP ERROR 404.

I am quite sure that any skilled 'web' programmer will fix this in microseconds. But I still have to study a lot...

May you help me (and every one else with the same issue)?

Thank you in advance.

Ciao

Alberto Utili

**Comment by fbarresi** [comment]
Hi Alberto!

I'm really sorry, I forgot to reply you. 😭  I will write a reply this evening!

Best regards,
Fede

**Comment by autili** [comment]
Hi Federico,
thank you very much.
In the meanwhile I am studying parts of your code.
Really interesting the possibility to expose
symbols also from .net ads servers.
Best regards
Alberto

**Comment by fbarresi** [comment, marked as answer]
Ok, Alberto!

Here a quick guide about how to run the software.

## Disclaimer
The software ist still under development: It may be buggy.
In case of errors please keep calm and open an issue in Github.

## Prerequisites
- dotnet core 3.1 sdk  ([download here](https://dotnet.microsoft.com/download/dotnet/3.1))
- Twincat ADS
- A Twincat ADS Client (of course I suggest you mine: [TwinCatAdsTool](https://github.com/fbarresi/TwinCatAdsTool))

## Build and run the code
1. Checkout the source code from github
2. Open a command prompt in your working directory
3. Navigate to the folder SoftBeckhoff with the command `cd SoftBeckhoff`
4. Build and Run the Software with: `dotnet run SoftBeckhoff.csproj --add-router=false --urls=http://localhost:5000`

Now you have SoftBeckhoff running on your PC! 🥳 

I'm assuming your are running the software locally, so you would not need a Twincat ADS Router.

## Setup a Symbol
You would see the API Addresses opened by the software in the console: `Now listening on: http://localhost:5000`
So in order to setup your symbols visit the swagger page on [http://localhost:5000/swagger](http://localhost:5000/swagger)
You can use the "try out" on every method  with `/symbols` and get, add, delete or update them. 
2 Symbols are always there for testing. 

## Start a Client
1. Start the TwinCatAdsTool
2. Set the PLC Port to 851 and Connect
3. You can now see under Variables the Symbols in your local ADS instance. 

I hope you can get started with this quick guide. 
If you still get stuck in problems don't hesitate to write!
